src/iterator.cr

```diff
@@ -297,30 +297,26 @@ module Iterator(T)
   # * If *reuse* is given, the array can be reused
   # * If *reuse* is an `Array`, this array will be reused
   # * If *reuse* is truthy, the method will create a new array and reuse it.
+  # * If *reuse* is falsey, the array will not be reused.
   #
   # This can be used to prevent many memory allocations when each slice of
   # interest is to be used in a read-only fashion.
   def cons(n : Int, reuse = false)
     raise ArgumentError.new "Invalid cons size: #{n}" if n <= 0
-    Cons(typeof(self), T, typeof(n)).new(self, n, reuse)
+    if reuse.nil? || reuse.is_a?(Bool)
+      Cons(typeof(self), T, typeof(n), Array(T)).new(self, n, Array(T).new(n), reuse)
+    else
+      Cons(typeof(self), T, typeof(n), typeof(reuse)).new(self, n, reuse, reuse)
+    end
   end
 
-  private struct Cons(I, T, N)
+  private struct Cons(I, T, N, V)
     include Iterator(Array(T))
     include IteratorWrapper
 
-    def initialize(@iterator : I, @n : N, reuse)
-      if reuse
-        if reuse.is_a?(Array)
-          @values = reuse
-        else
-          @values = Array(T).new(@n)
-        end
-        @reuse = true
-      else
-        @values = Array(T).new(@n)
-        @reuse = false
-      end
+    def initialize(@iterator : I, @n : N, values : V, reuse)
+      @values = values
+      @reuse = reuse ? true : false
```

Comment:
Consider `@reuse = !!reuse`, it seems to be a fairly common idiom, are there objections?

Reply:
No objections on my side, but we don't need to enforce it IMO.

Reply:
I have no opinion on the matter, except that I didn't like spending 5 lines on it which is what the original code looked like after my changes in the logic.